File: chart/src/main/java/cn/jingzhuan/lib/chart/data/BarDataSet.java
package cn.jingzhuan.lib.chart.data;

import android.graphics.Color;
import cn.jingzhuan.lib.chart.Viewport;
import cn.jingzhuan.lib.chart.component.HasValueYOffset;
import java.util.ArrayList;
import java.util.List;

import cn.jingzhuan.lib.chart.component.AxisY;
import cn.jingzhuan.lib.chart.component.AxisY.AxisDependency;


/**
 * Bar Data Set
 * Created by Donglua on 17/8/1.
 */

public class BarDataSet extends AbstractDataSet<BarValue> implements HasValueYOffset {

    private List<BarValue> mBarValues;
    private float mBarWidth = 20;
    private boolean mAutoBarWidth = false;
    private int mForceValueCount = -1;
    private float strokeThickness = 2;

    private float mBarWidthPercent = 0.8f;

    private boolean drawValueEnable = false;
    private int valueColor = Color.BLACK;
    private float valueTextSize = 24F;
    private ValueFormatter valueFormatter;

    public BarDataSet(List<BarValue> barValues) {
        this(barValues, AxisY.DEPENDENCY_BOTH);
    }

    public BarDataSet(List<BarValue> mBarValues, @AxisDependency int axisDependency) {
        this.mBarValues = mBarValues;
        setAxisDependency(axisDependency);
    }

    @Override
    public int getEntryCount() {
        if (mForceValueCount > 0) return mForceValueCount;

        if (mBarValues != null) {
            if (getMinValueCount() > mBarValues.size()) {
                return getMinValueCount();
            } else {
                return mBarValues.size();
            }
        }
        return 0;
    }

    @Override
    public void calcMinMax(Viewport viewport) {

        if (mBarValues == null || mBarValues.isEmpty())
            return;

        mViewportYMax = -Float.MAX_VALUE;
        mViewportYMin = Float.MAX_VALUE;

        for (BarValue e : getVisiblePoints(viewport)) {
            calcMinMaxY(e);
        }
        float range = mViewportYMax - mViewportYMin;
        if (Float.compare(getMinValueOffsetPercent(), 0f) > 0f) {
            mViewportYMin = mViewportYMin - range * getMinValueOffsetPercent();
        }
        if (Float.compare(getMaxValueOffsetPercent(), 0f) > 0f) {
            mViewportYMax = mViewportYMax + range * getMaxValueOffsetPercent();
        }
    }

    protected void calcMinMaxY(BarValue e) {

        if (e == null || !e.isEnable()) return;

        for (float v : e.getValues()) {
            if (!Float.isNaN(v) && !Float.isInfinite(v)) {
                mViewportYMin = Math.min(mViewportYMin, v);
                mViewportYMax = Math.max(mViewportYMax, v);
            }
        }
    }

    @Override
    public void setValues(List<BarValue> values) {
        this.mBarValues = values;
    }

    @Override
    public List<BarValue> getValues() {
        return mBarValues;
    }

    @Override
    public boolean addEntry(BarValue e) {
        if (e == null)
            return false;

        if (mBarValues == null) {
            mBarValues = new ArrayList<>();
        }

        calcMinMaxY(e);

        return mBarValues.add(e);
    }

    @Override
    public boolean removeEntry(BarValue e) {

        if (e == null) return false;

        calcMinMaxY(e);

        return mBarValues.remove(e);
    }

    @Override
    public int getEntryIndex(BarValue e) {
        return mBarValues.indexOf(e);
    }

    @Override
    public BarValue getEntryForIndex(int index) {
        return mBarValues.get(index);
    }

    public float getBarWidth() {
        return mBarWidth;
    }

    public void setBarWidth(float mBarWidth) {
        this.mBarWidth = mBarWidth;
    }

    public void setAutoBarWidth(boolean mAutoBarWidth) {
        this.mAutoBarWidth = mAutoBarWidth;
    }

    public boolean isAutoBarWidth() {
        return mAutoBarWidth;
    }

    public void setForceValueCount(int forceValueCount) {
        this.mForceValueCount = forceValueCount;
    }

    public int getForceValueCount() {
        return mForceValueCount;
    }

    public float getStrokeThickness() {
        return strokeThickness;
    }

    public void setStrokeThickness(float strokeThickness) {
        this.strokeThickness = strokeThickness;
    }

    @Override public float getMaxValueOffsetPercent() {
        return maxValueOffsetPercent;
    }

    @Override public float getMinValueOffsetPercent() {
        return minValueOffsetPercent;
    }

    @Override
    public void setMinValueOffsetPercent(float minValueOffsetPercent) {
        this.minValueOffsetPercent = minValueOffsetPercent;
    }

    @Override
    public void setMaxValueOffsetPercent(float maxValueOffsetPercent) {
        this.maxValueOffsetPercent = maxValueOffsetPercent;
    }

    public void setDrawValueEnable(boolean drawValueEnable) {
        this.drawValueEnable = drawValueEnable;
    }

    public boolean isDrawValueEnable() {
        return drawValueEnable;
    }

    public int getValueColor() {
        return valueColor;
    }

    public void setValueColor(int valueColor) {
        this.valueColor = valueColor;
    }

    public float getValueTextSize() {
        return valueTextSize;
    }

    public void setValueTextSize(float valueTextSize) {
        this.valueTextSize = valueTextSize;
    }

    public ValueFormatter getValueFormatter() {
        return valueFormatter;
    }

    public void setValueFormatter(ValueFormatter valueFormatter) {
        this.valueFormatter = valueFormatter;
    }

    public void setBarWidthPercent(float mBarWidthPercent) {
        this.mBarWidthPercent = mBarWidthPercent;
    }

    public float getBarWidthPercent() {
        return mBarWidthPercent;
    }
}

File: chart/src/main/java/cn/jingzhuan/lib/chart3/data/dataset/BarDataSet.kt
package cn.jingzhuan.lib.chart3.data.dataset

import android.graphics.Color
import cn.jingzhuan.lib.chart3.Viewport
import cn.jingzhuan.lib.chart3.axis.AxisY
import cn.jingzhuan.lib.chart3.axis.AxisY.AxisDependency
import cn.jingzhuan.lib.chart3.data.value.BarValue
import cn.jingzhuan.lib.chart3.formatter.IValueFormatter

import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt

/**
 * @since 2023-09-05
 * created by lei
 */
open class BarDataSet @JvmOverloads constructor(
    barValues: List<BarValue>,
    @AxisDependency axisDependency: Int = AxisY.DEPENDENCY_BOTH
) : AbstractDataSet<BarValue>(barValues, axisDependency) {

    var barWidth = 20f

    var minBarWidth = 0.0f

    var isAutoBarWidth = false

    var strokeThickness = 2f

    var barWidthPercent = 0.8f

    var isDrawValueEnable = false

    var valueColor = Color.BLACK

    var valueTextSize = 24f

    var valueFormatter: IValueFormatter? = null

    override fun getEntryCount(): Int {
        return if (forceValueCount > 0) forceValueCount else max(minValueCount, values.size)
    }

    override fun calcMinMax(viewport: Viewport) {
        if (values.isEmpty()) return
        maxVisibleY = -Double.MAX_VALUE
        minVisibleY = Double.MAX_VALUE

        val visiblePoints = getVisiblePoints(viewport)
        if (visiblePoints.isNullOrEmpty()) return

        for (value in visiblePoints) {
            calcMinMaxY(value)
        }

        val range = maxVisibleY - minVisibleY
        if (minValueOffsetPercent.compareTo(0f) > 0f) {
            minVisibleY -= range * minValueOffsetPercent
        }
        if (maxValueOffsetPercent.compareTo(0f) > 0f) {
            maxVisibleY += range * maxValueOffsetPercent
        }
    }

    override fun calcOverlayRatio(
        viewport: Viewport,
        baseDataSet: AbstractDataSet<*>
    ) : Float?{
        if (values.isEmpty()) return null
        val visiblePoints = getVisiblePoints(viewport)
        if (visiblePoints.isNullOrEmpty()) return null
        if (baseDataSet is CandlestickDataSet) {
            val foOpen = visiblePoints.first().values?.getOrNull(0) ?: return null
            val dataSize = baseDataSet.values.size
            val startIndex = max((dataSize * viewport.left).roundToInt(), 0)
            //k线在当前屏幕内的起点
            val fbOpen = baseDataSet.values.getOrNull(startIndex)?.open ?: return null
            if (!foOpen.isNaN() && foOpen != 0.0) {
                overLayRatio = (fbOpen / foOpen).toFloat()
                return overLayRatio
            }
        }
        return null
    }

    override fun calcOverlayMinMax(
        viewport: Viewport,
        ratio: Float?
    ) {
        if (values.isEmpty()) return
        maxVisibleY = -Double.MAX_VALUE
        minVisibleY = Double.MAX_VALUE

        val visiblePoints = getVisiblePoints(viewport)
        if (visiblePoints.isNullOrEmpty()) return
        overLayRatio = ratio

        for (value in visiblePoints) {
            calcMinMaxY(value)
        }

        val range = maxVisibleY - minVisibleY
        if (minValueOffsetPercent.compareTo(0f) > 0f) {
            minVisibleY -= range * minValueOffsetPercent
        }
        if (maxValueOffsetPercent.compareTo(0f) > 0f) {
            maxVisibleY += range * maxValueOffsetPercent
        }
    }

    protected fun calcMinMaxY(value: BarValue?) {
        if (value == null || !value.isEnable) return
        val values = value.values
        if (values != null && values.isNotEmpty()) {
            for (v in values) {
                if (v.isNaN()) continue
                if (v < minVisibleY) {
                    minVisibleY = v
                }
                if (v > maxVisibleY) {
                    maxVisibleY = v
                }
            }
        }
    }

    override fun addEntry(value: BarValue?): Boolean {
        if (value == null) return false
        calcMinMaxY(value)
        return values.add(value)
    }

    override fun removeEntry(value: BarValue?): Boolean {
        if (value == null) return false
        calcMinMaxY(value)
        return values.remove(value)
    }
}
